core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java

```diff
@@ -651,22 +609,57 @@ private void flattenProjection(RewriteRexShuttle shuttle,
         // NOTE jvs 10-Feb-2005:  This is a lame hack to keep special
         // functions which return row types working.
 
-        int j = 0;
         RexNode newExp = exp;
-        List<RexNode> oldOperands = ((RexCall) exp).getOperands();
-        if (oldOperands.get(0) instanceof RexInputRef) {
-          final RexInputRef inputRef = (RexInputRef) oldOperands.get(0);
-          final Ord<RelDataType> newField =
-              getNewFieldForOldInput(inputRef.getIndex());
-          newExp = rexBuilder.makeCall(exp.getType(),
-              ((RexCall) exp).getOperator(),
-              ImmutableList.of(rexBuilder.makeInputRef(newField.e, newField.i),
-                  oldOperands.get(1)));
-        }
-        for (RelDataTypeField field : newExp.getType().getFieldList()) {
-          flattenedExps.add(
-              Pair.of(rexBuilder.makeFieldAccess(newExp, field.getIndex()),
-                  fieldName + "$" + (j++)));
+        List<RexNode> operands = ((RexCall) exp).getOperands();
+
+        RexNode firstOp = operands.get(0);
+        RexNode secondOp = operands.get(1);
```

Comment:
It may be not safe to obtain the first operand here for the case when we have non-ITEM operator which has single operand.

Reply:
Fixed.